sw/device/lib/crypto/drivers/hmac.h

```diff
@@ -7,88 +7,176 @@
-  kHmacDigestNumBytes = kHmacDigestNumBits / 8,
-  /* Number of words in an HMAC or SHA-256 digest. */
-  kHmacDigestNumWords = kHmacDigestNumBytes / sizeof(uint32_t),
-  /* Number of bits in an HMAC key. */
-  kHmacKeyNumBits = 256,
-  /* Number of bytes in an HMAC key. */
-  kHmacKeyNumBytes = kHmacKeyNumBits / 8,
-  /* Number of words in an HMAC key. */
-  kHmacKeyNumWords = kHmacKeyNumBytes / sizeof(uint32_t),
+  /* Digest size for SHA-256/HMAC-256 in bits, bytes and word respectively. */
+  kHmacSha256DigestBits = 256,
+  kHmacSha256DigestBytes = kHmacSha256DigestBits / 8,
+  kHmacSha256DigestWords = kHmacSha256DigestBytes / sizeof(uint32_t),
+  /* Digest size for SHA-384/HMAC-384 in bits, bytes and word respectively. */
+  kHmacSha384DigestBits = 384,
+  kHmacSha384DigestBytes = kHmacSha384DigestBits / 8,
+  kHmacSha384DigestWords = kHmacSha384DigestBytes / sizeof(uint32_t),
+  /* Digest size for SHA-512/HMAC-512 in bits, bytes and word respectively. */
+  kHmacSha512DigestBits = 512,
+  kHmacSha512DigestBytes = kHmacSha512DigestBits / 8,
+  kHmacSha512DigestWords = kHmacSha512DigestBytes / sizeof(uint32_t),
+  /* Maximum key size supported by HW natively in bits and words respectively.
+   */
+  kHmacMaxKeyBits = 1024,
+  kHmacMaxKeyWords = kHmacMaxKeyBits / sizeof(uint32_t) / 8,
+  /* Maximum digest size supported by HW natively in bits and words
+     respectively. */
+  kHmacMaxDigestBits = 512,
+  kHmacMaxDigestWords = kHmacMaxDigestBits / sizeof(uint32_t) / 8,
+  /* Maximum internal block size bits, bytes and words respectively. */
+  kHmacMaxBlockBits = 1024,
+  kHmacMaxBlockBytes = kHmacMaxBlockBits / 8,
+  kHmacMaxBlockWords = kHmacMaxBlockBytes / sizeof(uint32_t),
 };
 
 /**
- * A typed representation of the HMAC or SHA256 digest.
+ * A typed representation of the SHA-2/HMAC digests.
  */
 typedef struct hmac_digest {
-  uint32_t digest[kHmacDigestNumWords];
+  uint32_t digest[kHmacMaxDigestWords];
+  // Length of `digest` in bytes.
+  size_t len;
 } hmac_digest_t;
 
 /**
- * A typed representation of an HMAC key.
+ * A typed representation of an HMAC key. HWIP supports 128, 256, 384, 512, 1024
+ * bit keys natively.
  */
 typedef struct hmac_key {
-  uint32_t key[kHmacKeyNumWords];
+  uint32_t key[kHmacMaxKeyWords];
+  // Length of `key` in bytes.
+  size_t len;
 } hmac_key_t;
 
 /**
- * Initializes the HMAC block in SHA256 mode.
- *
- * If the HMAC block is non-idle, calling this function aborts the previous
- * operation.
- *
- * This function resets the HMAC module to clear the digest register.
- * It then configures the HMAC block in SHA256 mode with little endian
- * data input and digest output.
+ * A context struct maintained for streaming operations.
  */
-void hmac_sha_init(void);
+typedef struct hmac_ctx {
+  // Back up cfg register, except SHA bit.
+  uint32_t cfg_reg;
+  // Need to keep some extra info around to reconfigure HW every time.
+  uint32_t key[kHmacMaxKeyWords];
+  // Length of `key` in words.
```

Comment:
nit: Clarify that `key_len == 0` to denote SHA2 mode of operation. You may want to consider adding a multi bit flag to denote HMAC enabled to simplify hardening.

```c
hardened_bool_t hmac_mode_en;
```

Reply:
In the hardening direction, I think there could be couple of improvements related to `hmac_ctx_t`. I think it would make sense to have 32b encoded field to denote which `hmac_mode` is running and use it to also recalculate values such as `msg_block_len` and `digest_len`. There is also another non-hardened flag `hw_started`.

For the moment, I would like to keep things simple so that we can focus on debugging RTL/driver issues. I have added this to the TODO items: #23191. I updated the comment line to emphasize that.